src/tests/Common/Directory.Build.targets

```diff
@@ -34,7 +35,47 @@
   </ItemDefinitionGroup>
 
   <Target Name="CopyDependencyToCoreRoot"
-          DependsOnTargets="ResolveAssemblyReferences;ResolveRuntimeFilesFromLocalBuild">
+    DependsOnTargets="ResolveAssemblyReferences;ResolveRuntimeFilesFromLocalBuild">
+
+    <!-- Copy apphost to crossgen2_publish directory.
+        The default configuration between product and tests are flipped. When we build project like:
+        `build -c debug -rc checked`, the corresponding test command is `src/tests/build -checked -p:LibrariesConfiguration=debug`,
+        instead of `-debug -p:RuntimeConfiguration=checked`. That forces us to either pass the
+        `HostConfiguration=debug` explicitly or fix this disparity; both of which will break the dev workflow.
+        
+        As a workaround, we will first check if the directory pointed by `HostConfiguration` exists, then check
+        `LibrariesConfiguration`.
+    -->
+
+    <PropertyGroup Condition="'$(Crossgen2Supported)' == 'true'">
+      <_targetOS>$(TargetOS)</_targetOS>
+      <_targetOS Condition="'$(_targetOS)' == 'windows'">win</_targetOS>
+      <_apphostPath Condition="Exists('$(ArtifactsBinDir)$(_targetOS)-$(TargetArchitecture).$(HostConfiguration)\corehost\apphost$(ExeSuffix)')">$(ArtifactsBinDir)$(_targetOS)-$(TargetArchitecture).$(HostConfiguration)\corehost\apphost$(ExeSuffix)</_apphostPath>
+      <_apphostPath Condition="'$(_apphostPath)' == '' and Exists('$(ArtifactsBinDir)$(_targetOS)-$(TargetArchitecture).$(LibrariesConfiguration)\corehost\apphost$(ExeSuffix)')">$(ArtifactsBinDir)$(_targetOS)-$(TargetArchitecture).$(LibrariesConfiguration)\corehost\apphost$(ExeSuffix)</_apphostPath>
+      <_toolsConfiguration Condition="Exists('$(ArtifactsBinDir)ILLink.Tasks\$(ToolsConfiguration)\$(NetCoreAppToolCurrent)\ILLink.Tasks.dll')">$(ToolsConfiguration)</_toolsConfiguration>
+      <_toolsConfiguration Condition="Exists('$(ArtifactsBinDir)ILLink.Tasks\$(LibrariesConfiguration)\$(NetCoreAppToolCurrent)\ILLink.Tasks.dll')">$(LibrariesConfiguration)</_toolsConfiguration>
+    </PropertyGroup>
+
+    <MakeDir Condition="'$(Crossgen2Supported)' == 'true'" Directories="$(ArtifactsObjDir)coreclr\crossgen2_publish\$(TargetOS).$(TargetArchitecture).$(RuntimeConfiguration)" />
+
+    <Copy
+      SourceFiles="$(_apphostPath)"
+      DestinationFiles="$(ArtifactsObjDir)coreclr\crossgen2_publish\$(TargetOS).$(TargetArchitecture).$(RuntimeConfiguration)\apphost$(ExeSuffix)"
+      Condition="'$(Crossgen2Supported)' == 'true'" />
```

Comment:
Instead of copying this into place, could we instead pass `HostConfiguration` as a property below like `ToolsConfiguration` with a value that will resolve to an apphost that has been built?

Reply:
I tried that, it wasn’t working. It’s basically just this file `apphost.exe`, which is resolved by some ResolveFrameworkReferences type of target that a `Copy` task later in this file requires and complains about. I haven’t intensively debugged it but I think there maybe a better solution which we can find later on.

Funny thing is crossgen2 publishing is either using the singlefilehost or AOT, so so this apphost requirement is just a bogus formality to satisfy a validation check.